firehose/main.go

```diff
@@ -0,0 +1,185 @@
+	outputName := argParser.String("n", "name", &argparse.Options{Required: false, Help: "unique name to use for this firehose, will be used to register a limacharlie.io Output if specified, otherwise assumes Output is already taken care of."})
+	argInvestigationID := argParser.String("i", "investigation-id", &argparse.Options{Required: false, Help: "firehose should only receive events marked with this investigation id."})
+	argTag := argParser.String("t", "tag", &argparse.Options{Required: false, Help: "firehose should only receive events from sensors tagged with this tag."})
+	argCategory := argParser.String("c", "category", &argparse.Options{Required: false, Help: "firehose should only receive detections from this category."})
+	// argSensorID := argParser.String("s", "sensor_id", &argparse.Options{Required: false, Help: "firehose should only receive detections and events from this sensor."})
+
+	err := argParser.Parse(os.Args)
+	if err != nil {
+		fmt.Print(argParser.Usage(err))
+		return lc.ClientOptions{}, lc.FirehoseOptions{}, lc.FirehoseOutputOptions{}, err
+	}
+
+	oid := ""
+	if argOID != nil {
+		oid = *argOID
+	}
+	listenOnIP, listenOnPort := getIPPort(*argListenOn)
+	destinationIP := listenOnIP
+	destinationPort := listenOnPort
+	if argDestination != nil && *argDestination != "" {
+		destinationIP, destinationPort = getIPPort(*argDestination)
+	}
+
+	isDeleteOnFailure := true
+	return lc.ClientOptions{
+			OID: oid,
+		},
+		lc.FirehoseOptions{
+			ListenOnIP:    listenOnIP,
+			ListenOnPort:  listenOnPort,
+			ConnectToIP:   destinationIP,
+			ConnectToPort: destinationPort,
+		},
+		lc.FirehoseOutputOptions{
+			UniqueName:        *outputName,
+			Type:              *lc.ParseOutputType(*argOutputType),
+			InvestigationID:   argInvestigationID,
+			Tag:               argTag,
+			Category:          argCategory,
+			IsDeleteOnFailure: &isDeleteOnFailure,
+		},
+		nil
+}
+
+func consumeMessages(fh lc.Firehose) {
+	for !fh.IsRunning() {
+		if len(fh.Messages) == 0 {
+			time.Sleep(1 * time.Second)
+		} else {
+			message := <-fh.Messages
+			log.Info().Msg(message.Message)
+		}
+	}
+}
+
+func consumeDroppedMessages(fh lc.Firehose) {
+	for !fh.IsRunning() {
+		if len(fh.Messages) == 0 {
+			time.Sleep(1 * time.Second)
+		} else {
+			errorMessage := <-fh.ErrorMessages
+			log.Error().Msg(fmt.Sprintf("Error processing: '%s'", errorMessage))
+		}
+	}
+}
+
+func main() {
+	log.Logger = log.Output(zerolog.ConsoleWriter{Out: os.Stderr})
+	interruptChannel := make(chan os.Signal, 1)
+	signal.Notify(interruptChannel, os.Interrupt)
+
+	clientOpts, fhOpts, fhOutputOpts, err := parseArgs()
+	if err != nil {
+		log.Err(err).Msg("error parsing arguments")
+		return
+	}
+
+	fmt.Println("Enter secret API key: ")
+	bytesAPIKey, err := terminal.ReadPassword(int(syscall.Stdin))
+	if err != nil {
+		log.Err(err).Msg("could not read API key")
+	}
+	clientOpts.APIKey = string(bytesAPIKey)
+
+	org, err := lc.MakeOrganization(clientOpts)
```

Comment:
This is another minor thing, but usually by convention constructors use the `New*` naming, so like `NewOrganization`.

Reply:
I thought the idiom was `NewX` would return a `*X` and `MakeX` would return a `X`

Reply:
Ah interesting, TIL. 👍